### PR TITLE
Fix a bug with transactions that have >10mb of attachments

### PR DESCRIPTION
**Pull request by mikehearn**
A user has encountered a bug whereby our limit of 10mb per transaction and per attachment is not being respected due to our use of batching in the fetch flows.  There should be no such limit on attachment sizes (or rather it should be a lot higher) but we put one in place to temporarily avoid upstream Artemis large message handling bugs. Once those are fixed we can re-batch this code if we want, and allow >10mb attachments.

This PR has a few commits:

* 3 which make some minor test utility changes and doc improvements.
* 1 which de-batches the test flow, and makes a new integration test pass. The integration test reproduces the bug reported to us by this user.

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/Utils.kt [hunk 1]
```diff
@@ -288,37 +288,39 @@ fun extractZipFile(inputStream: InputStream, toDirectory: Path) {
...
-fun getInputStreamAndHashFromOutputStream(baos: ByteArrayOutputStream): InputStreamAndHash {
-    // TODO: Consider converting OutputStream to InputStream without creating a ByteArray, probably using piped streams.
+fun ByteArrayOutputStream.getInputStreamAndHash(baos: ByteArrayOutputStream): InputStreamAndHash {
     val bytes = baos.toByteArray()
-    // TODO: Consider calculating sha256 on the fly using a DigestInputStream.
     return InputStreamAndHash(ByteArrayInputStream(bytes), bytes.sha256())
 }
 
-data class InputStreamAndHash(val inputStream: InputStream, val sha256: SecureHash.SHA256)
+data class InputStreamAndHash(val inputStream: InputStream, val sha256: SecureHash.SHA256) {
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/Utils.kt]: Can you move this to CoreTestUtils

Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/Utils.kt]: Seems it's not possible. That would move it to the test-utils module, which depends on node, but the RPC smoke tests need this class and the comments in gradle say smoke-tests aren't supposed to have any node code on the classpath.

Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/Utils.kt]: Seems the outcome of the discussion was to move it to test-common ... that doesn't have any kotlin code currently so it's a bigger change. Can I do it in a separate PR? It'll also make backporting it harder if it's done in this commit.

I did re-jig the class a bit to make it more Java friendly. PTAL?

File: core/src/main/kotlin/net/corda/core/flows/FetchAttachmentsFlow.kt [hunk 2]
```diff
@@ -15,8 +15,7 @@ import net.corda.core.serialization.SerializeAsTokenContext
  */
 @InitiatingFlow
 class FetchAttachmentsFlow(requests: Set<SecureHash>,
-                           otherSide: Party) : FetchDataFlow<Attachment, ByteArray>(requests, otherSide) {
-
+                           otherSide: Party) : FetchDataFlow<Attachment, ByteArray>(requests, otherSide, ByteArray::class.java) {
```
Comment (adagys) [inline core/src/main/kotlin/net/corda/core/flows/FetchAttachmentsFlow.kt]: Having to in effect specify ``ByteArray`` twice seems a bit superfluous, but it's probably ok for a temp change (I can see why that's needed)

File: node/src/integration-test/kotlin/net/corda/node/services/statemachine/LargeTransactionsTest.kt [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+ * Check that we can add lots of large attachments to a transaction and that it works OK, e.g. does not hit the
+ * transaction size limit (which should only consider the hashes).
+ */
+class LargeTransactionsTest {
+    companion object {
+        lateinit var hash1: SecureHash
+        lateinit var hash2: SecureHash
+        lateinit var hash3: SecureHash
+        lateinit var hash4: SecureHash
+    }
```
Comment (adagys) [inline node/src/integration-test/kotlin/net/corda/node/services/statemachine/LargeTransactionsTest.kt]: these could probably just be local vals and you could pass a list to the SendLargeTxFlow

File: core/src/main/kotlin/net/corda/core/Utils.kt [hunk 4]
```diff
@@ -291,37 +291,39 @@ fun extractZipFile(inputStream: InputStream, toDirectory: Path) {
...
+    companion object {
+        /**
+         * Get a valid InputStream from an in-memory zip as required for some tests. The zip consists of a single file
+         * called "z" that contains the given content byte repeated the given number of times.
+         * Note that a slightly bigger than numOfExpectedBytes size is expected.
+         */
+        @Throws(IllegalArgumentException::class)
+        @JvmStatic
+        fun createInMemoryTestZip(numOfExpectedBytes: Int, content: Byte): InputStreamAndHash {
```
Comment (adagys) [inline core/src/main/kotlin/net/corda/core/Utils.kt]: content could default to some random value?

File: core/src/main/kotlin/net/corda/core/flows/FetchDataFlow.kt [hunk 5]
```diff
@@ -54,12 +55,25 @@ abstract class FetchDataFlow<T : NamedByHash, in W : Any>(
...
+            val maybeItems = ArrayList<W>(toFetch.size)
+            send(otherSide, Request(toFetch))
+            for (hash in toFetch) {
+                // We skip the validation here (with unwrap { it }) because we will do it below in validateFetchResponse.
+                // The only thing checked is the object type.
+                maybeItems += receive(wrapperType, otherSide).unwrap { it }
+            }
             // Check for a buggy/malicious peer answering with something that we didn't ask for.
-            val downloaded = validateFetchResponse(maybeItems, toFetch)
+            val downloaded = validateFetchResponse(UntrustworthyData(maybeItems), toFetch)
```
Comment (adagys) [inline core/src/main/kotlin/net/corda/core/flows/FetchDataFlow.kt]: in theory we could receive items out of order, and that would cause an error in `validateFetchResponse`

File: core/src/main/kotlin/net/corda/core/flows/FetchAttachmentsFlow.kt (hunk 2 above)
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/flows/FetchAttachmentsFlow.kt]: It's not ideal but that's type erasure for you ...

File: core/src/main/kotlin/net/corda/core/flows/FetchDataFlow.kt (hunk 5 above)
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/flows/FetchDataFlow.kt]: I think we can just explicitly make that a protocol violation - how about I add a comment?

File: core/src/main/kotlin/net/corda/core/Utils.kt (hunk 4 above)
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/Utils.kt]: I'd rather leave it explicit. Consider if the default was random and I used it here. Then occasionally there'd be a collision and the hashes would not be different. This would cause the test to flake.

File: core/src/main/kotlin/net/corda/core/flows/FetchDataFlow.kt (hunk 5 above)
Comment (adagys) [inline core/src/main/kotlin/net/corda/core/flows/FetchDataFlow.kt]: actually, just realised that Artemis does preserve message order so it's not an issue in this case

File: node/src/integration-test/kotlin/net/corda/node/services/statemachine/LargeTransactionsTest.kt (hunk 3 above)
Comment (mikehearn) [inline node/src/integration-test/kotlin/net/corda/node/services/statemachine/LargeTransactionsTest.kt]: Good point! Done!

File: core/src/main/kotlin/net/corda/core/Utils.kt (hunk 4 above)
Comment (adagys) [inline core/src/main/kotlin/net/corda/core/Utils.kt]: But a collision seems extremely unlikely here if we use say random longs

**Comment by adagys** [comment]
Since Mike's away, I'll merge this PR if that's ok with you @shamsasari
